meson.build

```diff
@@ -3,13 +3,17 @@ project('ml-api', 'c', 'cpp',
   license: ['Apache-2.0'],
   meson_version: '>=0.50.0',
   default_options: [
+    'b_asneeded=false',
     'werror=true',
     'warning_level=1',
     'c_std=gnu89',
     'cpp_std=c++14'
   ]
 )
 
+add_global_link_arguments('-Wl,--no-as-needed', language: 'c')
+add_global_link_arguments('-Wl,--no-as-needed', language: 'cpp')
```

Comment:
If you already added `b_asneeded=false` up above I wonder why you *also* need to specify it as a link argument?

Also, either way you probably want `add_project_link_arguments()` rather than the global version.

Reply:
At least in `gcc version 9.3.0 (Ubuntu 9.3.0-11ubuntu0~18.04.1) ` removing `as_needed` is not enough as gcc implictly add `--as-needed` option.

without these lines,

build.ninja is compiled to

```
build c/libcapi-nnstreamer.so.1.8.0: c_LINKER  | c/libcapi-nnstreamer-pipeline.so.1.8.0.p/libcapi-nnstreamer-pipeline.so.1.8.0.symbols c/libcapi-nnstreamer-single.so.1.8.0.p/libcapi-nnstreamer-single.so.1.8.0.symbols
 LINK_ARGS = --verbose -Wl,--verbose -Wl,--no-undefined -shared -fPIC -Wl,--start-group -Wl,-soname,libcapi-nnstreamer.so.1 '-Wl,-rpath,$$ORIGIN/' -Wl,-rpath-link,/data/nnstreamer-api/build/c c/libcapi-nnstreamer-single.so.1.8.0 c/libcapi-nnstreamer-pipeline.so.1.8.0 -Wl,--end-group
```

However, when I add manually add `--verbose, -Wl--verbose` args to build.ninja to see what's actually going on, I was able to observe that `--as-needed` is attached by default.

```
COLLECT_GCC_OPTIONS='-o' 'c/libcapi-nnstreamer.so.1.8.0' '-v' '-shared' '-fPIC' '-mtune=generic' '-march=x86-64'
 /usr/lib/gcc/x86_64-linux-gnu/9/collect2 -plugin /usr/lib/gcc/x86_64-linux-gnu/9/liblto_plugin.so -plugin-opt=/usr/lib/gcc/x86_64-linux-gnu/9/lto-wrapper -plugin-opt=-fresolution=/tmp/ccdZpJ9f.res -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s --build-id --eh-frame-hdr -m elf_x86_64 --hash-style=gnu --as-needed -shared -z relro -o c/libcapi-nnstreamer.so.1.8.0 /usr/lib/gcc/x86_64-linux-gnu/9/../../../x86_64-linux-gnu/crti.o /usr/lib/gcc/x86_64-linux-gnu/9/crtbeginS.o -L/usr/lib/gcc/x86_64-linux-gnu/9 -L/usr/lib/gcc/x86_64-linux-gnu/9/../../../x86_64-linux-gnu -L/usr/lib/gcc/x86_64-linux-gnu/9/../../../../lib -L/lib/x86_64-linux-gnu -L/lib/../lib -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib -L/usr/lib/gcc/x86_64-linux-gnu/9/../../.. --verbose --no-undefined --start-group -soname libcapi-nnstreamer.so.1 -rpath $ORIGIN/ -rpath-link /data/nnstreamer-api/build/c c/libcapi-nnstreamer-single.so.1.8.0 c/libcapi-nnstreamer-pipeline.so.1.8.0 --end-group -lgcc --push-state --as-needed -lgcc_s --pop-state -lc -lgcc --push-state --as-needed -lgcc_s --pop-state /usr/lib/gcc/x86_64-linux-gnu/9/crtendS.o /usr/lib/gcc/x86_64-linux-gnu/9/../../../x86_64-linux-gnu/crtn.o
```

Thus, adding `-Wl,--no-asneeded` is also needed at least in my work environment.

Reply:
I agree `add_project_link_arguments()` is a better option, I'll make a patch for this.

Reply:
Yeah, interestingly enough it seems like some distros (Debian!) actually patch GCC in order to add --as-needed as an internal linker option, while it would seem to make more sense to enable that in packaging LDFLAGS...